linera-service/src/linera/main.rs

```diff
@@ -806,31 +816,118 @@ impl Runnable for Job {
                     .make_nodes(committee)?
                     .map(|(_, node)| node)
                     .collect::<Vec<_>>();
-                let blocks_infos_iter = blocks_infos.iter();
-                if bps.is_some() {
-                    let blocks_infos_iter = blocks_infos_iter.cycle();
-                    context
-                        .run_benchmark(
-                            bps,
-                            blocks_infos_iter,
-                            clients,
-                            transactions_per_block,
-                            epoch,
-                            chain_clients,
-                        )
-                        .await?;
-                } else {
-                    context
-                        .run_benchmark(
-                            bps,
-                            blocks_infos_iter,
-                            clients,
-                            transactions_per_block,
-                            epoch,
-                            chain_clients,
+
+                let shutdown_notifier = CancellationToken::new();
+                tokio::spawn(listen_for_shutdown_signals(shutdown_notifier.clone()));
+
+                // The bps control task will control the BPS from the threads. `crossbeam_channel` is used
+                // for two reasons:
+                // 1. it allows bounded channels with zero sized buffers.
+                // 2. it blocks the current thread if the message can't be sent, which is exactly
+                //    what we want to happen here. `tokio::sync::mpsc` doesn't do that. `std::sync::mpsc`
+                //    does, but it is slower than `crossbeam_channel`.
+                // Number 1 is the main reason. `tokio::sync::mpsc` doesn't allow 0 sized buffers.
+                // With a channel with a buffer of size 1 or larger, even if we have already reached
+                // the desired BPS, the tasks would continue sending block proposals until the channel's
+                // buffer is filled, which would cause us to not properly control the BPS rate.
+                let (sender, receiver) = crossbeam_channel::bounded(0);
+                let bps_control_task = tokio::spawn(async move {
+                    let mut recv_count = 0;
+                    let mut start = time::Instant::now();
+                    while let Ok(()) = receiver.recv() {
+                        recv_count += 1;
+                        if recv_count == num_chains {
+                            let elapsed = start.elapsed();
+                            if bps.is_some() {
```

Comment:
```suggestion
                            if let Some(bps) = bps {
```

Then we don't need the `unwrap`s.